sharedbuffers/mapped_struct.py

```diff
@@ -1425,21 +1576,52 @@ def proxied_list_richcmp(a, b, op):
     else:
         raise NotImplementedError
 
+def smallest_diff_key(A, B):
+    """return the smallest key adiff in A such that adiff not in B or A[adiff] != B[bdiff]"""
+    diff_keys = [k for k in A if k not in B or A[k] != B[k]]
+    return min(diff_keys)
+
+def dict_cmp_py3(A, B):
+    if len(A) != len(B):
+        return cmp(len(A), len(B))
+    try:
+        adiff = smallest_diff_key(A, B)
+    except ValueError:
+        # No difference.
+        return 0
+    bdiff = smallest_diff_key(B, A)
+    if adiff != bdiff:
+        return cmp(adiff, bdiff)
+    return cmp(A[adiff], B[bdiff])
+
+def dict_cmp(a, b):
+    if python2:
+        return cmp(a,b)
+    else:
+        return dict_cmp_py3(a,b)
+
 #@cython.ccall
 #@cython.returns(int)
 def proxied_list_cmp(a, b):
 
     alen = len(a)
     blen = len(b)
 
-    for i in xrange(min(alen, blen)):
+    for i in range(min(alen, blen)):
         selfe = a[i]
         othere = b[i]
 
-        if selfe < othere:
-            return -1
-        elif selfe > othere:
-            return 1
+        if isinstance(selfe, dict):
+            r = dict_cmp(selfe, othere)
+            if r:
+                return r
+        else:
+            if selfe is None and othere is None:
```

Comment:
```suggestion
            if selfe is othere:
```

It's a bit more general, and if they're the same object they're equal for sure. Might even help performance.